etna/packages/etna-py/src/mountetna/magma.py

```diff
@@ -287,7 +298,83 @@ def consume_pages():
 
     batch_size = 300
 
-    def update(self, update: UpdateRequest):
+    def _paged_update(self, update: UpdateRequest, page_size:int):
```

Comment:
This code is a bit tricky to follow. Perhaps it would be useful to:

- add some comments
- separate out some functionality, such as encapsulating the gathering all records to be updated and performing the actual updates
- refactoring how `results` is used. Maybe can be clarified with some renaming? For instance it is passed into `update_page()`, which is then fed into `add_to_page()`, but now `results` is treated as a `page` parameter?  Why are we re-adding to the page after we've performed the update?


The way I was originally trying to think about this code was:

1.  I have some records I want to update via pagination
2. I need to gather those records
3. I will then start to chunk those records and then send them to magma
4. I will collect the results of the updates (success/failures)
5. I will continue to paginate until I am done
6. Return the results

There are some elements of this, but I it could be improved?

I could be missing something though.

Reply:
Ok, refactored a bunch. I also realized that as written the code would not deal correctly with tables, which are the main thing making this update difficult.

Basically, there are two ways to attach data in Magma. One is to set the `parent` attribute on a record, the other is to set the reciprocal `table` or `collection` attribute on the parent record. In the case of a `table` attribute, this latter kind of update will replace any existing table contents. In such a case, the update will look like: `{ parent_model: { parent_record: { table_att: [ table_record1, table_record2 ] }, table_model: { table_record1: {}, table_record2 }` - so, to attach successfully in these cases the update must include all of the table records at once. Thus, we have to ensure that all of a table falls in a single "page" of our update. In practice this means we can't really guarantee that there are `page_size` records in the page, since a table could have any number of records, so we simply do our best (i.e. stop when the page count is exceeded). Therefore, when we flatten, we group tables with their parent records and only paginate over the parents.

Reply:
ahh i see

Reply:
code looks cleaner!